enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+  - No configuration should be required outside of this scope
+  - This enforces consistency across MachineSets, all Machines a MachinSet creates will either be Spot/Preemptible or On-Demand
+
+- A Machine should be paired 1:1 with an instance on the cloud provider
+  - If the instance is preempted/terminated, the cloud actuator should not replace it
+  - If the instance is preempted/terminated, the cloud provider should not replace it
+
+- The actuator is responsible for creation of the instance only and should not attempt to remediate problems
+
+- The actuator should not attempt to verify that an instance can be created before attempting to create the instance
+  - If the cloud provider does not have capacity, the Machine Health Checker can (given required MHC) remove the Machine after a period.
+    MachineSet will ensure the correct number of Machines are created.
+
+- There’s a [2 hours time frame](https://github.com/openshift/cluster-machine-approver/blob/master/csr_check.go#L38) in OpenShift for a machine to become a node.
+  If the Spot request is not satisfied during this time frame, it is assumed that the instance will never become a node.
+  This machine would be “garbage collected” by a given MHC.
+
+
+### Implementation Details
+
+#### Cloud Provider Implementation Specifics
+
+##### AWS
+
+###### Launching instances
+
+To launch an instance as a Spot instance on AWS, a [SpotMarketOptions](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotMarketOptions)
+needs to be added to the `RunInstancesInput`. Within this there are 3 options that matter:
+
+- InstanceInterruptionBehaviour (default: terminate): This must be set to `terminate` otherwise the SpotInstanceType cannot be `one-time`
+
+- SpotInstanceType (default: one-time): This must be set to `one-time` to ensure that each Machine only creates on EC2 instance and that the spot request is
+
+- MaxPrice (default: On-Demand price): This can be **optionally** set to a string representation of the hourly maximum spot price.
+  If not set, the option will default to the On-Demand price of the EC2 instance type requested
+
+The only option from this that needs exposing to the user from this is the `MaxPrice`, this option should be in an optional struct, if the struct is not nil,
+then spot instances should be used, if the MaxPrice is set, this should be used instead of the default On-Demand price.
+
+```
+type SpotMarketOptions struct {
+  MaxPrice *string `json:”maxPrice,omitempty”`
+}
+
+type AWSMachineProviderConfig struct {
+  ...
+
+  SpotMarketOptions *SpotMarketOptions `json:”spotMarketOptions,omitempty”`
+}
+```
+
+###### Termination Notices
+Termination notices on AWS are provided via the EC2 metadata service up to 2 minutes before an instance is due to be preempted.
+A DaemonSet should be deployed on the cluster which runs Pods on nodes labelled `machine.openshift.io/terminate-on-preemption: true`.
```

Comment:
Can we specify:
- What would run this daemonSet, I'd assume mao.
- Where does this controller implementation lives, I'd assume a new controller in each provider actuator repository. This binary could then be run by mao from the provider image similar to https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/operator.go#L284 and https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L225-L237

Reply:
I'm a little uncertain on how this would work at the moment, my current thinking is:
- New controller code for each termination controller goes into the respective actuator repositories
- This makes a new binary which goes into each mapi provider image for the cloud provider
- MAO syncs the daemonset in https://github.com/openshift/machine-api-operator/blob/e55d2b984245439082c3cd8f212b8c9c3380bbbe/pkg/operator/sync.go#L64-L74 if it is supported by the provider
- daemonset set's entrypoint to the second binary in the image that has been added

Does that make sense?